frontends/p4/specialize.h

```diff
@@ -113,6 +117,42 @@ class FindSpecializations : public Inspector {
     void postorder(const IR::Declaration_Instance* decl) override;
 };
 
+/** @brief Specializes each Parser and Control by substituting type arguments
+ * and constructor parameters.
+ *
+ * Specifically, each instantiation of each Parser (or Control) with
+ * constructor parameters will cause
+ *  - the creation of a new Parser (Control) with the actual arguments
+ *    substituted for the constructor parameters, and
+ *  - the replacement of the instantiation with a new instantiation using the
+ *    newly-specialized Parser (or Control).
+ *
+ * Note that this pass handles type substition for instantiating generic Parser
+ * or Control types, which is an experimental feature.
+ * 
+ * For example:
+ *
+```
+control c<T>(in T data)(bit<32> size) { ... }
+c<bit<32>>(16) c_inst;
+```
+ *
+ * is converted to
+ *
+```
+control cspec(in bit<32> data) { ... }
+cspec() c_inst;
+```
+ *
+ * A different specialization is made for each constructor invocation
+ * and Declaration_Instance.
+ *
+ * @pre Actual arguments are suitable for substitution, meaning that it must
+ * run after ConstantFolding and SimplifyExpressions.
+ *
+ * @post No instantiations of Parsers or Controls with constructor parameters
```

Comment:
Actually this may be incorrect; there won't remain such instantiations only after SpecializeAll.
We need to iterate because we specialize from the top-down: we need to know the arguments to specialize, so the arguments must be compile-time constants.